### DISCUSSION TITLE
[이진] assignment 기능구현

### DISCUSSION THREAD

**Discussion by im-binary** [Wanted-07-team-9/pre-onboarding-7th-3-1-9#15] [기능구현]
# 배포링크
[🪄배포링크](https://pre-onboarding-7th-3-1-9.vercel.app/)

## 사용 Tools

- TypeScript
- emotion

# 질환명 검색시 API 호출 통해서 검색어 추천 기능 구현

<img width="686" alt="스크린샷 2022-11-10 오후 10 35 35" src="https://user-images.githubusercontent.com/76990149/201105693-2a2dc1c6-4f9e-49c9-9c1d-d447fd398396.png">

- **사용자가 입력한 텍스트와 일치하는 부분 볼드 처리**

[👉 ParseBoldString.tsx](https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/ijin/src/utils/ParseBoldString.tsx)
```js
export const ParseBoldString = ({ sickName, keyword }: { sickName: string; keyword: string }) => {
  const result = sickName.match(new RegExp(keyword, 'i'));

  const beforeBoldText = sickName.slice(0, result.index);
  const boldText = sickName.slice(result.index, result.index + keyword.length);
  const afterBoldText = sickName.slice(result.index + keyword.length);

  return (
    <p
      css={css`
        padding-left: 24px;
      `}
    >
      {beforeBoldText}
      <strong>{boldText}</strong>
      {afterBoldText}
    </p>
  );
};
```

1. `match` 함수를 사용하여 문자열에 입력한 키워드와 같은 부분이 있는 index를 찾는다.
2. 그 index를 기준으로 문자열을 자른다.
4. `<strong>` 요소로 bold 처리할 문자열을 감싼다.

- **검색어가 없을 시 “검색어 없음” 표출**

<img width="648" alt="스크린샷 2022-11-10 오후 10 45 45" src="https://user-images.githubusercontent.com/76990149/201107987-3efdada8-1751-4ef8-8040-b35f9425b459.png">

# API 호출 최적화

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/a4989dfce4c50dd7b37fabd124ef22a32d8ceb1f/src/utils/cacheStorage.ts#L1-L61

- `cacheStorage` 라는 함수를 만들었습니다.
입력할 때마다 요청을 보내지 않으려면 이전에 요청했던 키워드에 대한 데이터를 어딘가에 저장해야 된다고 생각했습니다.

로컬스토리지는 키당 5MB를 저장할 수 있습니다. 용량이 작지도 않고, 캐시스토리지를 사용하면 비동기 처리도 해주어야 하기 때문에 복잡하다는 생각이 들어, 쉽게 문제를 해결하기 위해 데이터를 저장하는 공간으로 컬스토리지를 사용하기로 했습니다.

그래서 만료일을 따져 데이터를 로컬스토리지에 set, get, remove 하는 함수를 만들었습니다.

1.  데이터 요청을 보낼 때의 url을 로컬스토리지에 키로 저장하고 응답 데이터와 만료일을 값으로 저장했습니다. 
2. 또 같은 요청을 보내면, 로컬스토리지에 해당 요청 url의 이름을 가진 키가 있는지 확인하고 있으면 저장된 데이터를 반환해줍니다.
3. 만약 로컬스토리지에 해당 요청 url의 이름을 가진 키가 없다면 저장합니다.
4. 만약 로컬스토리지에 해당 요청 url의 이름을 가진 키가 있지만, 만료되었다면 지우고 다시 저장합니다.

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/a4989dfce4c50dd7b37fabd124ef22a32d8ceb1f/src/apis/fetchSickSearchList.ts#L4-L27

api 요청 보내는 함수에서 이 요청에 대해 캐싱을 할 건지 하지 않을 것인지 인자로 받을 수 있습니다. 기본값으로 캐싱을 하지 않도록 false로 지정하였고, 만료일도 기본값으로 30초로 지정해두었습니다.

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/a4989dfce4c50dd7b37fabd124ef22a32d8ceb1f/src/pages/SickSearchPage.tsx#L59-L63

실제 요청하는 곳에서 `{ useCache: true }` 로 넘겨주면 요청이 캐싱이 됩니다. 

만료일도 설정해주고 싶다면  `{ useCache: true, expireSeconds: (원하는만료일: number) }` 이런 식으로 입력해주면 됩니다.
(1을 입력하면 1초, 60을 입력하면 60초입니다. 초 단위로 입력하면 됩니다.)

![11111](https://user-images.githubusercontent.com/76990149/201122227-5bac600c-e21b-4018-bf0d-0ba20d717ac1.gif)

이전과 같은 검색어 입력 시 같은 요청이 가지 않습니당

# 키보드만으로 추천 검색어들로 이동 가능하도록 구현

```js
 const [selectIndex, setSelectIndex] = useState<number>(-1);
```

선택된 검색어의 index를 저장하는 상태를 하나 만들었습니다.
기본값이 -1인 이유는 검색어 배열의 index의 첫 index 값은 0이기 때문입니다.

```js
const onKeyUp = (e: KeyboardEvent<HTMLInputElement>) => {
    if (searchKeyword == null) {
      return;
    }

    const maxIndex = sickSearchList.length - 1;

    if (e.key === 'ArrowUp') {
      if (selectIndex === -1) {
        setSelectIndex(maxIndex);
      } else {
        setSelectIndex(selectIndex - 1);
      }
    }

    if (e.key === 'ArrowDown') {
      if (selectIndex === maxIndex) {
        setSelectIndex(-1);
      } else {
        setSelectIndex(selectIndex + 1);
      }
    }
  };
```

저는 keyUP 이벤트를 사용하여 키보드로 검색어를 이동할 수 있도록 했습니다.

1. 윗 키를 눌렀을 때 ⬆
  - 현재 선택된 index에서 1씩 감소시킵니다.
  - 만약 `selectIndex` (현재 선택된 index)가 -1일 경우 더 이상 올라갈 검색어가 없으니 맨 아래의 index를 `setSelectIndex(maxIndex)` 를 해줍니다.

3. 아래 키를 눌렀을 때 ⬇
  -  현재 선택된 index에서 1씩 증가시킵니다.
  - 만약 `selectIndex` (현재 선택된 index)가 맨 아래의 index일 경우 맨 위의 index, -1로 setSelectIndex(-1)을 해줍니다.

```js
<li css={sickItem} className={index === selectIndex ? 'select' : ''}></li>
```

검색어 배열의 index와 `selectIndex` (현재 선택된 index)가 같은지를 비교하여 요소에 `select` 라는 클래스명이 붙도록 했고 그에 따라 스타일을 지정해주었습니다.

![22222](https://user-images.githubusercontent.com/76990149/201123506-7207d09a-9f58-4b65-bc36-9f553f726169.gif)

**Comment by jun-05** [comment]
로컬스토리지로 캐싱기능 구현하신게 신기한거 같아요.
그런데 일부 특정 상황에서 API 호출이 잘 안되는거 같습니다
